### PR TITLE
.NET 9

### PR DESCRIPTION
**Pull request by chenzhitong**
update to .net 9

### CODE REVIEW

File: src/Neo.VM/ScriptBuilder.cs [hunk 1]
```diff
@@ -126,7 +126,7 @@ public ScriptBuilder EmitPush(bool value)
         /// <returns>A reference to this instance after the emit operation has completed.</returns>
         public ScriptBuilder EmitPush(ReadOnlySpan<byte> data)
         {
-            if (data == null)
```
Comment (shargon) [inline src/Neo.VM/ScriptBuilder.cs:129]: Data Length == 0 is empty too?

File: src/Plugins/TokensTracker/Extensions.cs [hunk 2]
```diff
@@ -28,7 +28,7 @@ public static bool NotNull(this StackItem item)
 
         public static string ToBase64(this ReadOnlySpan<byte> item)
         {
-            return item == null ? String.Empty : Convert.ToBase64String(item);
+            return item.IsEmpty ? String.Empty : Convert.ToBase64String(item);
```
Comment (Jim8y) [inline src/Plugins/TokensTracker/Extensions.cs:31]: is empty can be used on null as well?

File: tests/Neo.VM.Tests/UT_Utility.cs [hunk 3]
```diff
@@ -54,7 +54,7 @@ public void TestGetBitLength()
             var random = new Random();
 
             // Big Number (net standard didn't work)
-            VerifyGetBitLength(BigInteger.One << 32 << int.MaxValue, 2147483680);
+            Assert.ThrowsException<OverflowException>(() => VerifyGetBitLength(BigInteger.One << 32 << int.MaxValue, 2147483680));
```
Comment (Jim8y) [inline tests/Neo.VM.Tests/UT_Utility.cs:57]: why is this change? the behavor is no longer the same?

File: tests/Neo.Extensions.Tests/UT_ByteExtensions.cs [hunk 4]
```diff
@@ -25,7 +25,7 @@ public class UT_ByteExtensions
         public void TestToHexString()
         {
             byte[] nullStr = null;
-            Assert.ThrowsException<NullReferenceException>(() => nullStr.ToHexString());
+            Assert.ThrowsException<ArgumentNullException>(() => nullStr.ToHexString());
```
Comment (Jim8y) [inline tests/Neo.Extensions.Tests/UT_ByteExtensions.cs:28]: does this mean Net9.0 has changed the exception type?

Comment (chenzhitong) [reply tests/Neo.Extensions.Tests/UT_ByteExtensions.cs:28]: yes

File: tests/Neo.VM.Tests/UT_Utility.cs (hunk 3 above)
Comment (chenzhitong) [reply tests/Neo.VM.Tests/UT_Utility.cs:57]: `BigInteger.One << 32 << int.MaxValue` will throw OverflowException in .net 9

File: tests/Neo.Extensions.Tests/UT_ByteExtensions.cs (hunk 4 above)
Comment (nan01ab) [reply tests/Neo.Extensions.Tests/UT_ByteExtensions.cs:28]: > does this mean Net9.0 has changed the exception type?

`ToHexString` uses `Convert.ToHexStrongLower` when run with NET.9.
May need to add an argument null-check to keep same semantics.

File: tests/Neo.VM.Tests/UT_Utility.cs (hunk 3 above)
Comment (shargon) [reply tests/Neo.VM.Tests/UT_Utility.cs:57]: Then this test is useless

**Comment by cschuchardt88** [comment]
I rather not go to `dotnet` `9.0` it removes a lot of stuff and changes the way things work. and community is not to happy about it. Also its a `Short Term Release`.

**Comment by chenzhitong** [comment]
> I rather not go to `dotnet` `9.0` it removes a lot of stuff and changes the way things work. and community is not to happy about it. Also its a `Short Term Release`.
> 
> Also you missed `global.json`

I didn't miss global.json

```c#
{
  "sdk": {
    "version": "9.0.100",
    "rollForward": "latestFeature",
    "allowPrerelease": false
  }
}
```

I’m not sure whether to upgrade to .NET 9. I'll leave it up to you core developers to decide!

**Comment by vncoelho** [comment]
I am in favor, next year we move to long term again and so on.

**Comment by shargon** [comment]
> I am in favor, next year we move to long term again and so on.

me too

**Comment by cschuchardt88** [comment]
This will break the functionally of `RestServer`

**Comment by cschuchardt88** [comment]
> I’m not sure whether to upgrade to .NET 9. I'll leave it up to you core developers to decide!

No dont leave it up to them. If it's dotnet 9.0 then have to upgrade. there is no way around it.

**Review by cschuchardt88 — APPROVED** [review]

**Review by cschuchardt88 — COMMENTED** [review]
Why are you changing `*.cs` files?

**Comment by chenzhitong** [comment]
> Why are you changing `*.cs` files?

Because some of the exception types change, and some of the dotnet format rules change.

**Comment by shargon** [comment]
> > I’m not sure whether to upgrade to .NET 9. I'll leave it up to you core developers to decide!
> 
> No dont leave it up to them. If it's dotnet 9.0 then have to upgrade. there is no way around it.

That's true

File: src/Neo.VM/ScriptBuilder.cs [hunk 5]
```diff
@@ -126,7 +126,8 @@ public ScriptBuilder EmitPush(bool value)
...
+            ReadOnlySpan<byte> nullData = null;
```
Comment (shargon) [inline src/Neo.VM/ScriptBuilder.cs:129]: What is this?

File: src/Plugins/TokensTracker/Extensions.cs [hunk 6]
```diff
@@ -28,7 +29,8 @@ public static bool NotNull(this StackItem item)
 
         public static string ToBase64(this ReadOnlySpan<byte> item)
         {
-            return item == null ? String.Empty : Convert.ToBase64String(item);
+            ReadOnlySpan<byte> nullData = null;
```
Comment (shargon) [inline src/Plugins/TokensTracker/Extensions.cs:32]: ?

File: src/Neo.VM/ScriptBuilder.cs (hunk 5 above)
Comment (chenzhitong) [reply src/Neo.VM/ScriptBuilder.cs:129]: Because `if (data == null)` will fail the format check in .net 9.

File: src/Plugins/TokensTracker/Extensions.cs (hunk 6 above)
Comment (chenzhitong) [reply src/Plugins/TokensTracker/Extensions.cs:32]: Because `item == null` will fail the format check in .net 9.

Comment (shargon) [reply src/Plugins/TokensTracker/Extensions.cs:32]: We should add test to ensure if IsEmpty check null and not Length != 0

Comment (chenzhitong) [reply src/Plugins/TokensTracker/Extensions.cs:32]: Running dotnet format automatically changes `item == null` to `item.IsEmpty`. If you use `item.IsEmpty`, the unit test will fail, I tried it. So I've changed it to `ReadOnlySpan<byte> nullData = null; item == nullData;`

Comment (cschuchardt88) [reply src/Plugins/TokensTracker/Extensions.cs:32]: just edit the `.editorconfig` and turn that off.

File: .editorconfig [hunk 7]
```diff
@@ -26,6 +26,7 @@ indent_size = 4
 insert_final_newline = true
 trim_trailing_whitespace = true
 charset = utf-8
+dotnet_diagnostic.CA2265.severity = none
```
Comment (cschuchardt88) [inline .editorconfig:29]: this should be on `line 58`

File: tests/Neo.ConsoleService.Tests/Neo.ConsoleService.Tests.csproj [hunk 8]
```diff
@@ -10,9 +10,13 @@
...
-    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.10.0" />
-    <PackageReference Include="MSTest.TestAdapter" Version="3.5.0" />
-    <PackageReference Include="MSTest.TestFramework" Version="3.5.0" />
+    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.11.1" />
+    <PackageReference Include="MSTest.TestAdapter" Version="3.6.3" />
+    <PackageReference Include="MSTest.TestFramework" Version="3.6.3" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <PackageReference Update="FluentAssertions" Version="6.12.2" />
```
Comment (cschuchardt88) [inline tests/Neo.ConsoleService.Tests/Neo.ConsoleService.Tests.csproj:19]: Please use the `Directory.Build.props` in the directory to update this.

**Review by cschuchardt88 — APPROVED** [review]

**Comment by shargon** [comment]
I would like to reduce the changes before merge it, and test this changes with the proper ut

File: src/Neo.CLI/CLI/MainService.cs [hunk 9]
```diff
@@ -521,13 +521,13 @@ private void WriteBlocks(uint start, uint count, string path, bool writeStart)
...
                     fs.Seek(sizeof(uint), SeekOrigin.Begin);
-                    fs.Read(buffer, 0, buffer.Length);
+                    fs.ReadExactly(buffer);
                     start += BitConverter.ToUInt32(buffer, 0);
                     fs.Seek(sizeof(uint), SeekOrigin.Begin);
                 }
                 else
                 {
-                    fs.Read(buffer, 0, buffer.Length);
+                    fs.ReadExactly(buffer);
```
Comment (Jim8y) [inline src/Neo.CLI/CLI/MainService.cs:530]: this is the only place where code is changed in the core. @shargon

Comment (chenzhitong) [reply src/Neo.CLI/CLI/MainService.cs:530]: This paragraph was automatically changed after running `dotnet format` in dotnet 9, and I saw that the logic hadn't changed, so I submitted it.

**Review by Jim8y — DISMISSED** [review]

File: tests/Neo.VM.Tests/UT_ScriptBuilder.cs [hunk 10]
```diff
@@ -42,6 +42,23 @@ public void TestEmit()
...
+            using (ScriptBuilder script = new())
+            {
+                Assert.ThrowsException<ArgumentNullException>(() =>
+                {
+                    ReadOnlySpan<byte> span = null;
+                    script.EmitPush(span);
+                });
+
+                ReadOnlySpan<byte> span = [];
+                script.EmitPush(span);
```
[the remark this answers is not in the document]
Comment (Jim8y) [reply tests/Neo.VM.Tests/UT_ScriptBuilder.cs:57]: the warning shows `Comparing a span to 'null' might be redundant, the 'null' literal will be implicitly converted to a 'Span<T>.Empty'`

File: src/Neo.CLI/CLI/MainService.cs (hunk 9 above)
Comment (cschuchardt88) [reply src/Neo.CLI/CLI/MainService.cs:530]: turn it off in the `.editorconfig`

File: tests/Neo.VM.Tests/UT_ScriptBuilder.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (shargon) [reply tests/Neo.VM.Tests/UT_ScriptBuilder.cs:57]: If it worked like this in net8, we should remove the null check and push always empty array

[the remark this answers is not in the document]
Comment (shargon) [reply tests/Neo.VM.Tests/UT_ScriptBuilder.cs:57]: But previously seems that with null we receive an error and with empty an empty push, not we can't

[the remark this answers is not in the document]
Comment (Jim8y) [reply tests/Neo.VM.Tests/UT_ScriptBuilder.cs:57]: Tested new UT on master, it does not work as well

```
Test method Neo.Test.UT_ScriptBuilder.TestNullAndEmpty threw exception: 
System.ArgumentNullException: Value cannot be null. (Parameter 'data')
    at Neo.VM.ScriptBuilder.EmitPush(ReadOnlySpan`1 data) in C:\Users\liaoj\git\neo\src\Neo.VM\ScriptBuilder.cs:line 130
   at Neo.Test.UT_ScriptBuilder.TestNullAndEmpty() in C:\Users\liaoj\git\neo\tests\Neo.VM.Tests\UT_ScriptBuilder.cs:line 57
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodBaseInvoker.InvokeWithNoArgs(Object obj, BindingFlags invokeAttr)

```

[the remark this answers is not in the document]
Comment (shargon) [reply tests/Neo.VM.Tests/UT_ScriptBuilder.cs:57]: Then we should push always empty array and avoid the exception

File: .editorconfig [hunk 11]
```diff
@@ -57,7 +57,6 @@ indent_size = 2
 [*.{cs,vb}]
 # Use block-scoped namespace
 csharp_style_namespace_declarations = block_scoped:error
-dotnet_diagnostic.CA2265.severity = none
```
Comment (chenzhitong) [inline .editorconfig]: Why did you change this back?

File: tests/Neo.VM.Tests/UT_ScriptBuilder.cs [hunk 12]
```diff
@@ -42,6 +42,23 @@ public void TestEmit()
...
         }
 
+        [TestMethod]
+        public void TestNullAndEmpty()
+        {
+            using (ScriptBuilder script = new())
+            {
+                Assert.ThrowsException<ArgumentNullException>(() =>
+                {
+                    ReadOnlySpan<byte> span = null;
```
Comment (Jim8y) [inline tests/Neo.VM.Tests/UT_ScriptBuilder.cs:52]:
```suggestion
                    ReadOnlySpan<byte> span = [];
```

**Review by Jim8y — DISMISSED** [review]

**Review by shargon — APPROVED** [review]

**Review by shargon — APPROVED** [review]

File: tests/Neo.VM.Tests/UT_ScriptBuilder.cs [hunk 13]
```diff
@@ -42,6 +42,21 @@ public void TestEmit()
             }
         }
 
+        [TestMethod]
+        public void TestNullAndEmpty()
+        {
+            using (ScriptBuilder script = new())
+            {
+                ReadOnlySpan<byte> span = null;
```
Comment (cschuchardt88) [inline tests/Neo.VM.Tests/UT_ScriptBuilder.cs:50]: should be using `empty`, `null` doesn't really work for `spans`

File: src/Neo.VM/Types/Map.cs [hunk 14]
```diff
@@ -29,7 +28,7 @@ public class Map : CompoundType, IReadOnlyDictionary<PrimitiveType, StackItem>
         /// </summary>
         public const int MaxKeySize = 64;
 
-        private readonly OrderedDictionary<PrimitiveType, StackItem> dictionary = new();
+        private readonly Collections.OrderedDictionary<PrimitiveType, StackItem> dictionary = new();
```
Comment (cschuchardt88) [inline src/Neo.VM/Types/Map.cs:31]: Where is `Collections` pointing to; `system` or `vm`?

**Review by cschuchardt88 — APPROVED** [review]

File: src/Neo.VM/Types/Map.cs (hunk 14 above)
Comment (Jim8y) [reply src/Neo.VM/Types/Map.cs:31]: VM

**Comment by shargon** [comment]
We need to revert it https://github.com/neo-project/neo-devpack-dotnet/pull/1257#issuecomment-2507022502
